docs/authentication.md

```diff
@@ -0,0 +1,174 @@
+
+## Patterns
+
+The first step to identifying which authentication pattern you need is understanding the [data-fetching strategy](/docs/basic-features/data-fetching) you want. We can then determine which authentication providers support this strategy. There are two main strategies:
+
+- Use [static generation](/docs/basic-features/pages#static-generation-recommended) to [server-render](/docs/basic-features/pages#static-generation-without-data) a loading state, followed by fetching user data client-side.
+- Fetch user data server-side to eliminate a flash of unauthenticated content.
+
+### Static Generation
+
+Next.js automatically determines that a page is static if there are no blocking data requirements. This means the absence of `getServerSideProps` and `getInitialProps` in the page. Instead, your page can render a loading state from the server, followed by fetching the user client-side.
+
+One advantage of this pattern is it allows pages to be served from a global CDN and [preloaded](/docs/api-reference/next/link) using `<Link />`. Depending on the size of your application, [React hydration](https://reactjs.org/docs/react-dom.html#hydrate) could take some time. If you serve a loading shell from the server, this allows React to hydrate while you're fetching user data. In practice, this results in a faster TTI ([Time to Interactive](https://web.dev/interactive/)).
+
+Let's look at an example for a profile page. This will initially render a loading skeleton. Once the request for a user has finished, it will show the user's name.
+
+```jsx
+// pages/profile.js
+
+import useUser from '../lib/useUser'
+import Layout from '../components/Layout'
+
+export default Profile = () => {
+  // Fetch the user client-side
+  const { user } = useUser({ redirectTo: '/login' })
+
+  // Server-render loading state
+  if (!user || user.isLoggedIn === false) {
+    return <Layout>Loading...</Layout>
+  }
+
+  // Once the user request finishes, show the user
+  return (
+    <Layout>
+      <h1>Your Profile</h1>
+      <pre>{JSON.stringify(user, undefined, 2)}</pre>
+    </Layout>
+  )
+}
+```
+
+You can view this example in action [here](https://next-iron-session.now.sh/).
+
+### Server-side Rendering
+
+If you export an `async` function called `getServerSideProps` from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.
+
+```jsx
+export async function getServerSideProps(context) {
+  return {
+    props: {} // Will be passed to the page component as props
+  };
+}
+```
+
+Let's transform the profile example to use [server-side rendering](/docs/basic-features/pages#server-side-rendering). If there's a session, return `user` as a prop to the `Profile` component in the page. Notice there is not a loading skeleton in [this example](https://next-iron-session.now.sh/).
```

Comment:
I am glad you're linking to next-iron-session directly. But I feel it would be better to create your own example for Vercel so that if I make changes to next-iron-session examples, your documentation won't be in 404 or outdated code-wise..

Reply:
I want to add that I plan to keep this example up to date two months later. next-iron-session gained a good userbase, and I am still using it, so I'll maintain it as long as needed.

So I am fine if you keep it this way.